xmlua/c14n.lua

```diff
@@ -0,0 +1,257 @@
+    end,
+    [ffi.C.XML_PI_NODE] = function(document, raw_node)
+      return ProcessingInstruction.new(document, raw_node)
+    end,
+    [ffi.C.XML_COMMENT_NODE] = function(document, raw_node)
+      return Comment.new(document, raw_node)
+    end,
+    [ffi.C.XML_DOCUMENT_NODE] = function(document, raw_node)
+      return Document.new(raw_node)
+    end,
+    [ffi.C.XML_DOCUMENT_TYPE_NODE] = function(document, raw_node)
+      return DocumentType.new(document, raw_node)
+    end,
+    [ffi.C.XML_DOCUMENT_FRAG_NODE] = function(document, raw_node)
+      return DocumentFragment.new(document, raw_node)
+    end,
+    [ffi.C.XML_NOTATION_NODE] = function(document, raw_node)
+      return Notation.new(document, raw_node)
+    end,
+    [ffi.C.XML_HTML_DOCUMENT_NODE] = function(document, raw_node)
+      error("XML_HTML_DOCUMENT_NODE not implemented")     -- TODO: implement
+    end,
+    [ffi.C.XML_DTD_NODE] = function(document, raw_node)
+      error("XML_DTD_NODE not implemented")               -- TODO: implement
+    end,
+    [ffi.C.XML_ELEMENT_DECL] = function(document, raw_node)
+      return ElementDeclaration.new(document, raw_node)
+    end,
+    [ffi.C.XML_ATTRIBUTE_DECL] = function(document, raw_node)
+      return AttributeDeclaration.new(document, raw_node)
+    end,
+    [ffi.C.XML_ENTITY_DECL] = function(document, raw_node)
+      return EntityDeclaration.new(document, raw_node)
+    end,
+    [ffi.C.XML_NAMESPACE_DECL] = function(document, raw_node)
+      return NamespaceDeclaration.new(document, raw_node)
+    end,
+    [ffi.C.XML_XINCLUDE_START] = function(document, raw_node)
+      error("XML_XINCLUDE_START not implemented")         -- TODO: implement
+    end,
+    [ffi.C.XML_XINCLUDE_END] = function(document, raw_node)
+      error("XML_XINCLUDE_END not implemented")           -- TODO: implement
+    end,
+    [ffi.C.XML_DOCB_DOCUMENT_NODE] = function(document, raw_node)
+      error("XML_DOCB_DOCUMENT_NODE not implemented")     -- TODO: implement
+    end,
+  }, {
+    __index = function(self, key)
+      error("Unknown node type: " .. tostring(key))
+    end
+  })
+
+  function wrap_raw_node(document, raw_node)
+    if raw_node == ffi.NULL then
+      return nil
+    end
+    return type_generators[tonumber(raw_node.type)](document, raw_node)
+  end
+end
+
+
+
+--- Canonicalise an xmlDocument or set of elements.
+-- @param self xmlDoc from which to canonicalize elements
+-- @param callback function to determine if a node should be included in the canonicalized output.
+--        Signature: `boolean = function(node, parent)`
+-- @param mode any of C14N_1_0, C14N_EXCLUSIVE_1_0 (default), C14N_1_1
+-- @param inclusive_ns_prefixes array, or space-separated string, of namespace prefixes to include
+-- @param with_comments if truthy, comments will be included (default: false)
+-- @return string containing canonicalized xml
+function C14n:c14n_execute(callback, mode, inclusive_ns_prefixes, with_comments)
```

Comment:
What is the difference of `c14n_save()` and `c14n_execute()`?
Specifying target nodes as an array or specifying target nodes as a callback?
Can we provide one function that dispatches internally based on the given array or callback?

Reply:
in libxml2 the callbackased one is the basis. The other is build on top as a convenience. For me; if you have traversed the tree and know what you want to include use `save`, if you only know the keys you want to in/exclude but haven't traversed the tree, then the `execute` one makes more sense.

Reply:
OK. Can we provide only one method something like the following in XMLua?

```lua
document:canonicalize([node1, node2, ...], ...) -- Use xmlC14NDocSave() internally
document:canonicalize(function() ... end, ...) -- Use xmlC14NExecute() internally
```

Should we provide different methods in XMLua too?

Reply:
Keeping both, which then aligns with the underlying implementation and docs, makes it easier to grasp for users. My 2cts. But your choice in the end. Let me know if you want me to change it.

Reply:
XMLua provides user friendly API not one-to-one API for libxml2.
Let's choose user friendly API approach.

Reply:
Thanks.